src/topology/sheaves/Godement.lean

```diff
@@ -0,0 +1,124 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import topology.sheaves.sheaf
+import topology.sheaves.limits
+import topology.sheaves.skyscraper
+import topology.sheaves.stalks
+import category_theory.preadditive.injective
+
+/-!
+# Godement resolution
+
+For a presheaf `𝓕 : (opens X)ᵒᵖ ⥤ C`, we can embedded `𝓕` into a sheaf `∏ₓ skyscraper(𝓕ₓ)` where
+`x` ranges over `X` and `𝓕 ⟶ ∏ₓ skyscraper(𝓕ₓ)` is mono.
+
+## Main definition
+* `godement_presheaf`: for a presheaf `𝓕`, its Godement presheaf is `∏ₓ skyscraper(𝓕ₓ)`
+* `to_godement_presheaf`: the canonical map `𝓕 ⟶ godement_presheaf 𝓕` sending `s : 𝓕(U)` to a
+  bundle of stalks `x ↦ sₓ`.
+-/
+
+noncomputable theory
+
+section presheaf
+
+open Top
+open topological_space
+open category_theory
+open category_theory.limits
+
+universes u v
+
+variables {X : Top.{u}} {C : Type u} [category.{u} C]
+variables [has_limits C] [has_terminal C] [has_colimits C]
+variables [Π (x : X) (U : opens X), decidable (x ∈ U)]
+variables (𝓕 : presheaf C X)
+
+/--
+The `godement_presheaf` for a presheaf `𝓕` is defined as a product presheaf `∏ₓ skyscraper(𝓕ₓ)`
+-/
+def godement_presheaf : presheaf C X :=
+∏ (λ x, skyscraper_presheaf x (𝓕.stalk x) : X → presheaf C X)
+
+/--
+The sections of `godement_presheaf` on opens `U` is isomorphic to `∏ₓ skyscraper(x, 𝓕ₓ)(U)`, i.e.
+the categorical definition and the concrete definition agree.
+-/
+@[simps] def godement_presheaf_obj (U : (opens X)ᵒᵖ) :
+  (godement_presheaf 𝓕).obj U ≅ ∏ (λ x, (skyscraper_presheaf x (𝓕.stalk x)).obj U) :=
+limit_obj_iso_limit_comp_evaluation _ _ ≪≫
+{ hom := lim_map { app := λ _, 𝟙 _, naturality' := by { rintros ⟨x⟩ ⟨y⟩ ⟨⟨(rfl : x = y)⟩⟩, refl } },
+  inv := lim_map { app := λ _, 𝟙 _, naturality' := by { rintros ⟨x⟩ ⟨y⟩ ⟨⟨(rfl : x = y)⟩⟩, refl } },
+  hom_inv_id' :=
+  begin
+    ext,
+    erw [category.assoc, lim_map_π, ←category.assoc, lim_map_π, category.id_comp, category.comp_id,
+      category.comp_id],
+  end,
+  inv_hom_id' :=
+  begin
+    dsimp,
+    ext,
+    erw [category.assoc, lim_map_π, ←category.assoc, lim_map_π, category.comp_id, category.id_comp,
+      category.comp_id],
+  end }
```

Comment:
Maybe its better to prove this for arbitrary product of presheaves then specialise it

Reply:
I was curious about the failure of defeq and the necessity of `≪≫`, so I investigated and found that the following works. Even if we redefine `discrete.functor` using `eq_to_hom` it won't help with the defeq (still requires a [category_theory.eq_to_hom_app](https://leanprover-community.github.io/mathlib_docs/category_theory/eq_to_hom.html#category_theory.eq_to_hom_app)). But I do think your explicit `app := λ _, 𝟙 _,` is better. 
```
universes v u v₁ u₁

variables {C : Type u} [category.{v} C] {D : Type u₁} [category.{v₁} D]
  {ι : Type*} (f : ι → D ⥤ C) [has_products_of_shape ι C]

example (d : D) : (∏ f).obj d ≅ ∏ λ x, (f x).obj d :=
limit_obj_iso_limit_comp_evaluation _ _ ≪≫ eq_to_iso begin
  congr, apply functor.hext,
  { intro, refl },
  { intros X Y f, apply heq_of_eq,
    cases X, cases Y,
    obtain ⟨⟨rfl : X = Y⟩⟩ := f, refl },
end
```

Reply:
Okay, I think this should be the preferred definition:
```
example (d : D) : (∏ f).obj d ≅ ∏ λ x, (f x).obj d :=
limit_obj_iso_limit_comp_evaluation _ _ ≪≫ limits.lim.map_iso (discrete.nat_iso $ λ X, iso.refl _)
```